### PR TITLE
Fix multiple attributes support

### PR DESCRIPTION
**Pull request by Guiguiprim**
Improves `PathMap`:
- Add iterator over all values matching a path (allow multiple attributes, with the same path and with sub-paths)  #490 #502
- Keep insertion order to avoid warnings/errors https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#legacy-derive-helpers
- Correctly handle field attributes on one-of #507

### CODE REVIEW

File: prost-build/src/code_generator.rs [hunk 1]
```diff
@@ -251,26 +251,35 @@ impl<'a> CodeGenerator<'a> {
...
             .field_attributes
             .get_field(fq_message_name, field_name)
-            .cloned()
         {
-            self.push_indent();
-            self.buf.push_str(&attributes);
-            self.buf.push('\n');
+            for _ in 0..self.depth {
+                buf.push_str("    ");
+            }
```
Comment (LucioFranco) [inline prost-build/src/code_generator.rs:277]: Why not call `self.push_indent` here?

File: prost-build/src/code_generator.rs [hunk 2]
```diff
@@ -251,26 +251,35 @@ impl<'a> CodeGenerator<'a> {
...
+            buf.push_str(&attribute);
+            buf.push('\n');
+        }
+        if !buf.is_empty() {
```
Comment (LucioFranco) [inline prost-build/src/code_generator.rs:281]: Is this check needed? Appending a empty str should be a noop

File: prost-build/src/code_generator.rs [hunk 3]
```diff
@@ -251,26 +251,35 @@ impl<'a> CodeGenerator<'a> {
...
+            self.buf.push_str(&buf);
         }
     }
 
     fn append_field_attributes(&mut self, fq_message_name: &str, field_name: &str) {
         assert_eq!(b'.', fq_message_name.as_bytes()[0]);
-        // TODO: this clone is dirty, but expedious.
-        if let Some(attributes) = self
+        let mut buf = String::new();
```
Comment (LucioFranco) [inline prost-build/src/code_generator.rs:269]: What is the benefit of this extra String? I don't see why you can't just push directly into the buf.

File: prost-build/src/path.rs [hunk 4]
```diff
@@ -1,61 +1,107 @@
 //! Utilities for working with Protobuf paths.
 
-use std::collections::HashMap;
 use std::iter;
 
 /// Maps a fully-qualified Protobuf path to a value using path matchers.
 #[derive(Debug, Default)]
 pub(crate) struct PathMap<T> {
-    matchers: HashMap<String, T>,
+    // insertion order might actually matter (to avoid warning about legacy-derive-helpers)
```
Comment (LucioFranco) [inline prost-build/src/path.rs:8]: Nice catch 👍

File: prost-build/src/path.rs [hunk 5]
```diff
@@ -1,61 +1,107 @@
...
-            // Then, try matching path suffixes.
-            .chain(suffixes(fq_path))
-            // Then, try matching path prefixes.
-            .chain(prefixes(fq_path))
-            // Then, match the global path. This matcher must never fail, since the constructor
-            // initializes it.
-            .chain(iter::once("."))
-            .flat_map(|path| self.matchers.get(path))
-            .next()
+    pub(crate) fn get<'a>(&'a self, fq_path: &'_ str) -> Iter<'a, T> {
```
Comment (LucioFranco) [inline prost-build/src/path.rs:19]: I think you can elide this lifetime

`pub(crate) fn get(&self, fq_path: &str) -> Iter<'_, T>`

File: prost-build/src/path.rs [hunk 6]
```diff
@@ -1,61 +1,107 @@
...
+    pub(crate) fn get_field<'a>(&'a self, fq_path: &'_ str, field: &'_ str) -> Iter<'a, T> {
+        Iter::new(self, format!("{}.{}", fq_path, field))
+    }
+
+    #[allow(unused)]
+    pub(crate) fn get_best<'a>(&'a self, fq_path: &'_ str) -> Option<&'a T> {
+        self.find_best_matching(fq_path)
+    }
+
+    pub(crate) fn get_best_field<'a>(&'a self, fq_path: &'_ str, field: &'_ str) -> Option<&'a T> {
```
Comment (LucioFranco) [inline prost-build/src/path.rs:32]: would a better name here be `get_first_field()` and might be worth it to document this behavior.

File: prost-build/src/path.rs [hunk 7]
```diff
@@ -1,61 +1,107 @@
...
+                }
+                None => return None,
+            }
+        }
+    }
+}
+
+impl<'a, T> std::iter::FusedIterator for Iter<'a, T> {}
+
+/// Given a fully-qualified path, retruns a sequence of paths:
```
Comment (LucioFranco) [inline prost-build/src/path.rs:89]:
```suggestion
/// Given a fully-qualified path, returns a sequence of paths:
```

**Review by LucioFranco — CHANGES_REQUESTED** [review]
Overall this looks great, I left a few inline comments/questions.

File: prost-build/src/code_generator.rs (hunk 3 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:269]: The extra string is adressing the removed TODO.

It avoids the need for cloning the map (lifetime issue between `&self` for itering the map and `&mut self` for editing the buffer).

File: prost-build/src/code_generator.rs (hunk 2 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:281]: Probably not

File: prost-build/src/code_generator.rs (hunk 1 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:277]: Because I'm using a local buffer ...

File: prost-build/src/code_generator.rs (hunk 3 above)
Comment (LucioFranco) [reply prost-build/src/code_generator.rs:269]: Ah got it didn't catch that borrowing issue.

File: prost-build/src/code_generator.rs (hunk 1 above)
Comment (LucioFranco) [reply prost-build/src/code_generator.rs:277]: Right xD many buf variables

File: prost-build/src/code_generator.rs (hunk 3 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:269]: I improved it to remove the local buffer, only the call to `self.push_ident()` was problematic.

File: prost-build/src/code_generator.rs (hunk 2 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:281]: Removed

File: prost-build/src/code_generator.rs (hunk 1 above)
Comment (Guiguiprim) [reply prost-build/src/code_generator.rs:277]: Reworked.

File: prost-build/src/path.rs (hunk 4 above)
Comment (Guiguiprim) [reply prost-build/src/path.rs:8]: Done

File: prost-build/src/path.rs (hunk 6 above)
Comment (Guiguiprim) [reply prost-build/src/path.rs:32]: Done

File: prost-build/src/path.rs (hunk 5 above)
Comment (Guiguiprim) [reply prost-build/src/path.rs:19]: Done

File: prost-build/src/path.rs (hunk 7 above)
Comment (Guiguiprim) [reply prost-build/src/path.rs:89]: Done

File: prost-build/src/code_generator.rs [hunk 8]
```diff
@@ -38,6 +38,12 @@ pub struct CodeGenerator<'a> {
     buf: &'a mut String,
 }
 
+fn _push_indent(buf: &mut String, depth: u8) {
```
Comment (LucioFranco) [inline prost-build/src/code_generator.rs:41]: Not following why this has a `_`?

Comment (Guiguiprim) [reply prost-build/src/code_generator.rs]: Fixed

My C++ brain made me bug, I needed a local private free function which name would not conflict with the class method...

**Review by LucioFranco — APPROVED** [review]
Thanks!

**Comment by Logarithmus** [comment]
@LucioFranco new release maybe? We can't use this fix until it ends up in a new release.

**Comment by u0xy** [comment]
A new release would be awesome for this fix.

**Pull request merged**